# Assignment 4

Turn in the assignment via Canvas.

To write legible answers you will need to be familiar with both [Markdown](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet) and [Latex](https://www.latex-tutorial.com/tutorials/amsmath/)

Before you turn this problem in, make sure everything runs as expected. First, restart the kernel (in the menubar, select Runtime→→Restart runtime) and then run all cells (in the menubar, select Runtime→→Run All).

Make sure you fill in any place that says "YOUR CODE HERE" or "YOUR ANSWER HERE", as well as your name below:

In [ ]:
NAME = "Juan Lee"
STUDENT_ID = ""

# Question 1

Solve the folowing [8-puzzle](https://en.wikipedia.org/wiki/15_puzzle) using steepest ascent hillclimbing. 

![8puzzle.png](https://lh3.googleusercontent.com/cjob9NQOFnv80vfG71y2QBi7blTo9zl_UNcnFz1UhICK59Kc8zWBAfRa26yssTv-b7iQT2TiHlCaOLKPW-fXY_1TdIuT8Yzj9fhpKVwuW9cZSF8ehn0oTY5QA3jUfK-AMhrJu5s2aw=w2400)

1. State the heuristic used.
2. List the steps taken as comma seprated values. For example if the first three moves consist of moving the 3 tile, then the 8 tile, then the 6 tile. Your answer would start like: 3,8,6

Here is the description of the algorithm from the book:

![hill-climbing.png](https://lh3.googleusercontent.com/e1tTwqFAQvLGhfNdq50sQIzx2yxvP0N2Jnxh18qek59-SzrXaO4pUNnANUILUBd5qoPVoEHreBv8eo2f1zMKEtdLqS8xV47dNu-WNpXreqnEGNMCe7mih1AxG4yPdjA_gL6WjWr48w=w2400)

# [1. The heuristic that i used is number of puzzle pieces that are at their goal states if I were to swap the empty space with that tile. 
 So from the start, current's heauristic value would be 4.  Neighbor 4 would have heuristic value of 3. Since 4>3, back to current state. Neighbor 3 would have heauristic value of 5. 5>4 So next move would be swap spaces with neighbor 3. 
2. 3,2,1,8]

# Question 2

Consider the following constraint satisfaction problem. A linear graph has nodes of the following colors:

- Red
- Yellow
- Green
- Blue
- Violet

Each node has a domain of {1, 2, ..., 9}.

Each node type has the following constraints on its value:

- Red - No contraints
- Yellow - equals the rightmost digit of of the product of all its neighbors
- Green - equals the rightmost digit of the sum of all its neighbors 
- Blue - equals the leftmost digit of the sum of all its neighbors
- Violet - equals the leftmost digit of the product of all of its neighbors

As a reminder here is the pseudo code for the Min-Conflicts search algorithm:

![minconflicts](https://docs.google.com/drawings/d/e/2PACX-1vTIdOyAKDEoK6evNWQBkx9X5kl2I7GLaUkE9TdFDRqyyNFiHeFDrA-Sm7sLob2wMSzoBk_cliRhs8PY/pub?w=927&amp;h=474)

**Notes:**

- It's possible that you won't converge to a solution in a single run. Try a few runs to see if you get to a solution.
- The example is to show you what a problem looks like, we will test/grade your program on different examples


Complete the function *solve_csp* defined below. You may find some helper functions useful.

In [ ]:
from random import *
import math

#find the value that minimizes the conflict and returning that value
def conflict(nodes, asize, arc_n, node_val, i):
  value = 0
  if nodes[i] == 'Y':
    value = 1
    for p in range(len(arc_n[i])):
      value *= node_val[arc_n[i][p]]
    value = value%10
  elif nodes[i] == 'G':
    value = 0
    for p in range(len(arc_n[i])):
      value += node_val[arc_n[i][p]]
    value = value%10
  elif nodes[i] == 'B':
    value = 0
    for p in range(len(arc_n[i])):
      value += node_val[arc_n[i][p]]
    while value >= 10:
      value /= 10
  elif nodes[i] == 'V':
    value = 1
    for p in range(len(arc_n[i])):
      value *= node_val[arc_n[i][p]]
    while value >= 10:
      value /= 10
  elif nodes[i] == 'R':
    value = node_val[i]
  return math.floor(value)

# choosing the random conflicted variable and returning the index of it
def ran_conflict(nodes, asize, arc_n, node_val):
  found = False
  count = 0
  while found == False:
    index = randrange(0, asize)
    val = conflict(nodes, asize, arc_n, node_val, index)
    if val != node_val[index]:
      return index


def arc_organize(arcs, len, alen):
  #making a two demensional array to store the arcs
  arc_neighbor = []
  for i in range(len):
    arc_neighbor.append([])

  #stores the values of the given arcs into the 2d array
  for p in range(alen):
    arc_neighbor[arcs[p][0]].append(arcs[p][1])
    arc_neighbor[arcs[p][1]].append(arcs[p][0])
  return arc_neighbor

#checks if a solution has been found to this csp problem
def solution(nodes, asize, arc_n, node_val):
   for i in range(asize):
     val = conflict(nodes, asize, arc_n, node_val, i)
     if val != node_val[i]:
       return False
   return True

def solve_csp(nodes, arcs, max_steps):
    """
    This function solves the csp using the MinConflicts Search
    Algorithm.

    INPUTS:
    nodes:      a list of letters that indicates what type of node it is,
                the index of the node in the list indicates its id
                letters = {R, Y, G, B, V}
                
    arcs:       a list of tuples that contains two numbers, indicating the 
                IDS of the nodes the arc connects. 
                
    max_steps:  max number of steps to make before giving up

    RETURNS: a list of values for the solution to the CSP where the 
             index of the value corresponds the value for that
             given node.
    """
    #creating a random initial csp state with random values from the range of 0 to 9
    node_values = []
    asize = len(nodes)
    for i in range(asize):
      node_values.append(randrange(1, 10))
    #Creating a 2D array for keeping track of neighbors
    organized_arc = arc_organize(arcs, asize, len(arcs))

    val = 0
    index = randrange(0, asize)
    # Min-conflicts algorithm 
    for i in range(max_steps):
      if solution(nodes, asize, organized_arc, node_values):
        return node_values
      index = ran_conflict(nodes, asize, organized_arc, node_values)
      result = conflict(nodes, asize, organized_arc, node_values, index)
      if result != 0:
        node_values[index] = result

    return node_values

### Test Cases

Below we've included 4 test cases to help you debug your code. Your submitted code will be tested on different cases than these, but if your implementation of the above Min-Conflicts search algorithm pseudocode is able to solve these problems, you should be good to go.

In [ ]:
# Test Case 1

nodes = 'YGVRB'
arcs = [(0,1), (0,2), (1,2), (1,3), (1,4), (2,3), (2,4)]
max_steps = 1000

for _ in range(max_steps):
    sol = solve_csp(nodes, arcs, max_steps)
    if sol != []:
        break
        
all_solutions = [[1, 1, 1, 7, 2],[2, 1, 2, 4, 3],[2, 6, 7, 6, 1],[2, 8, 9, 6, 1],
                 [3, 3, 1, 5, 4],[6, 2, 8, 7, 1],[6, 7, 8, 2, 1],[6, 9, 4, 8, 1]]

if sol == []:
    print('No solution')
else:
    if sol in all_solutions:
        print('Solution found:', sol)
    else:
        print('ERROR: False solution found:', sol)

ERROR: False solution found: [5, 5, 1, 8, 6]


In [ ]:
# Test Case 2

nodes = 'YVBGR'
arcs = [(0,1), (0,2), (1,3), (2,4)]
max_steps = 1000

for _ in range(max_steps):
    sol = solve_csp(nodes, arcs, max_steps)
    if sol != []:
        print(nodes)
        break
        
all_solutions = [[1, 1, 1, 1, 9], [1, 3, 7, 3, 6], [1, 7, 3, 7, 2], [1, 9, 9, 9, 8]]

if sol == []:
    print('No solution')
else:
    if sol in all_solutions:
        print('Solution found:', sol)
    else:
        print('ERROR: False solution found:', sol)

YVBGR
ERROR: False solution found: [1, 2, 5, 2, 4]


In [ ]:
# Test Case 3

nodes = 'VYGBR'
arcs = [(0,1), (1,2), (2,3), (3,4)]
max_steps = 1000

for _ in range(max_steps):
    sol = solve_csp(nodes, arcs, max_steps)
    if sol != []:
        print(nodes)
        break
        
all_solutions = [[2, 2, 1, 9, 8],[3, 3, 1, 8, 7],[4, 4, 1, 7, 6],[5, 5, 1, 6, 5],[5, 5, 3, 8, 5],
                 [6, 6, 1, 5, 4],[7, 7, 1, 4, 3],[8, 8, 1, 3, 2],[8, 8, 6, 8, 2],[9, 9, 1, 2, 1]]

if sol == []:
    print('No solution')
else:
    if sol in all_solutions:
        print('Solution found:', sol)
    else:
        print('ERROR: False solution found:', sol)

VYGBR
Solution found: [6, 6, 1, 5, 4]


In [ ]:
# Test Case 4

nodes = 'YGVBR'
arcs = [(0,1), (0,2), (1,3), (2,3), (3,4), (1,2)]
max_steps = 1000

for _ in range(max_steps):
    sol = solve_csp(nodes, arcs, max_steps)
    if sol != []:
        print(nodes)
        break
        
all_solutions = [[4, 4, 1, 9, 4],[4, 7, 2, 1, 1],[4, 7, 2, 1, 2],[4, 7, 2, 1, 3],[4, 7, 2, 1, 4],
                 [4, 7, 2, 1, 5],[4, 7, 2, 1, 6],[4, 7, 2, 1, 7],[4, 7, 2, 1, 8],[4, 7, 2, 1, 9],
                 [4, 8, 3, 1, 1],[4, 8, 3, 1, 2],[4, 8, 3, 1, 3],[4, 8, 3, 1, 4],[4, 8, 3, 1, 5],
                 [4, 8, 3, 1, 6],[4, 8, 3, 1, 7],[4, 8, 3, 1, 8],[5, 1, 5, 1, 4],[5, 1, 5, 1, 5],
                 [5, 1, 5, 1, 6],[5, 1, 5, 1, 7],[5, 1, 5, 1, 8],[5, 1, 5, 1, 9]]

if sol == []:
    print('No solution')
else:
    if sol in all_solutions:
        print('Solution found:', sol)
    else:
        print('ERROR: False solution found:', sol)

YGVBR
Solution found: [4, 7, 2, 1, 3]


# Question 3   -  the n rooks problem

Rooks can move any number of squares horizontally or vertically on a chess board. The n rooks problem is to arrange rooks on an $n\times n$ board in such a way that none of the rooks could bump into another by making any of its possible horizontal or vertical moves. 

For this problem, the variables are each column (labeled 0, 1, ... , $ n-1$), the the domain consists of each possible row (also labeled 0, 1, ... , $ n-1$). In each column we place a rook on row 0, row 1, ... , row $n-1$.

For example, if $n=2$ we have only two solutions to this problem:

    R @
    @ R 
    
or

    @ R
    R @ 
   
How many possible solutions are there to this CSP, in terms of n?  
Also, give a simple proof that your answer is correct.

[There are n! possible solutions to this CSP. For example, for when n = 1, there is only one possible solution for the rook and 1! = 1. For when n = 2, we have just discussed that there are 2 possible solution and 2! = 2*1 = 2. For when n = 3, there are 6 possible solutions and 3! = 3 * 2 * 1 = 6. The 6 possible solution are:

    R @ @
    @ R @
    @ @ R

    @ @ R
    @ R @
    R @ @

    @ R @
    R @ @
    @ @ R

    @ @ R
    R @ @
    @ R @

    R @ @
    @ @ R
    @ R @

    @ R @
    @ @ R
    R @ @

]



### Python Library to solve CSPs

One useful Python module for solving these types of problems is called *constraint*. In the next cell you'll see how to load this module into a Jupyter notebook running in Colab. 

We'll use this module to solve the following simple CSP. Suppose our variables are x and y, and the values they are allowed to assume are numbers in the domain 

            {1, 2, ... , 100}, 

and with constraints be that $x^2=y$ and that $x$ is odd.

In [ ]:
# The following line imports the constraint module into a Jupyter notebook running in Colab.
# (This only needs to be done once per session)
!pip install python-constraint

# Let's load all the functions available in this module
# (This only needs to be done once per session)
from constraint import *

# Now we initialize a new problem and add the two variables.
problem = Problem()
problem.addVariables(["x", "y"], list(range(1, 101)))

# So far there are no constraints.  If we solve the problem with no constraints, 
# then every possible combination of x and y values from 1 to 100 are valid solutions.
solutions = problem.getSolutions()

# There are 100*100 = 10000 solutions in there:
print( len(solutions) )

10000


In [ ]:
# Add the given constraints:
problem.addConstraint(lambda a, b: a**2 == b and (a % 2) == 1, ("x", "y"))

# There are only 5 solutions that satisfy those constraints:
solutions = problem.getSolutions()
print( len(solutions) )

5


In [ ]:
# Here are all 5 solutions:
print(solutions)

[{'x': 9, 'y': 81}, {'x': 7, 'y': 49}, {'x': 5, 'y': 25}, {'x': 3, 'y': 9}, {'x': 1, 'y': 1}]


Notice that the solutions consist of a list of dictionaries, where each dictionary represents a solution. For example, the first solution is {'x': 9, 'y': 81}, since with $x=9$ and $y=81$ it's true that $x^2=y$.

### The n-rooks problem revisited

Modify the example before to solve the n rooks problem. You only need one call to *problem.addConstraint*, as before, since you can use logical 'and's to include as many other constraints in the lambda function as you need.

In [ ]:
# YOUR CODE HERE

#setting up the problem
problem = Problem()
#for part b change the n value to 11
n = 8
#initializing the variables
cols = range(n)
rows = range(n)
problem.addVariables(cols, rows)

#adding the constraints
for column1 in cols:
     for column2 in cols:
         if column1 < column2:
             problem.addConstraint(lambda row1, row2: row1 != row2,(column1, column2))
#getting and printing out the number of solutions for the n rook problem
solutions = problem.getSolutions()
print(len(solutions))

40320



(a) How many ways are there of arranging 8 rooks on an 8x8 board so that none impede the others?

(b) How many ways are there arranging 11 rooks on an 11x11 board so that none impede the others?

(a) [40320 different ways of arrange the rooks when n = 8. 8! = 40320]

(b) [39916800 different ways of arranging the rooks when n = 11. 11! = 39916800]